plugins/parquet-store/parquet_store.cpp

```diff
@@ -0,0 +1,564 @@
+      auto list_array = std::static_pointer_cast<arrow::ListArray>(array);
+      if (auto fixed_array = map_array(lt.value_type(), list_array->values()))
+        return std::make_shared<arrow::ListArray>(
+          lt.to_arrow_type(), list_array->length(), list_array->value_offsets(),
+          *fixed_array, list_array->null_bitmap(), list_array->null_count());
+      return {};
+    },
+    [&](const map_type& mt) -> std::optional<std::shared_ptr<arrow::Array>> {
+      auto ma = std::static_pointer_cast<arrow::MapArray>(array);
+      auto key_array = map_array(mt.key_type(), ma->keys());
+      auto val_array = map_array(mt.value_type(), ma->items());
+      if (!key_array && !val_array)
+        return {};
+      auto ka = key_array ? *key_array : ma->keys();
+      auto va = val_array ? *val_array : ma->items();
+      return std::make_shared<arrow::MapArray>(mt.to_arrow_type(), ma->length(),
+                                               ma->value_offsets(), ka, va,
+                                               ma->null_bitmap(),
+                                               ma->null_count());
+    },
+    [&](const record_type& rt) -> std::optional<std::shared_ptr<arrow::Array>> {
+      auto struct_array = std::static_pointer_cast<arrow::StructArray>(array);
+      // TODO: detail::zip doesn't seem to work with a generator and a vector
+      auto it = struct_array->fields().begin();
+      auto children = std::vector<std::shared_ptr<arrow::Array>>{};
+      children.reserve(rt.num_fields());
+      auto modified = false;
+      for (const auto& field : rt.fields()) {
+        auto mapped_arr = map_array(field.type, *it);
+        if (mapped_arr) {
+          modified = true;
+          children.push_back(*mapped_arr);
+        } else {
+          children.push_back(*it);
+        }
+        ++it;
+      }
+      if (!modified)
+        return {};
+      return std::make_shared<arrow::StructArray>(
+        rt.to_arrow_type(), struct_array->length(), children,
+        struct_array->null_bitmap(), struct_array->null_count());
+    },
+    [&](const auto&) -> std::optional<std::shared_ptr<arrow::Array>> {
+      return {};
+    },
+  };
+  return caf::visit(f, t);
+}
+
+/// Transform a given `ChunkedArray` according to the provided VAST type
+/// `ChunkedArray`s only occurr at the outermost level, and the VAST type
+/// that is not properly represented at this level is `enumeration_type`.
+std::optional<std::shared_ptr<arrow::ChunkedArray>>
+restore_enum_chunk_array(const vast::type& t,
+                         std::shared_ptr<arrow::ChunkedArray> array) {
+  auto f = detail::overload{
+    [&](const enumeration_type& et)
+      -> std::optional<std::shared_ptr<arrow::ChunkedArray>> {
+      return map_chunked_array(et, array, fix_enum_array);
+    },
+    [&](
+      const list_type&) -> std::optional<std::shared_ptr<arrow::ChunkedArray>> {
+      return map_chunked_array(t, array, map_array);
```

Comment:
I don't understand why you're passing in `t` rather than the already resolved `list_type` here. If that is passed to `map_array` wouldn't you just need to visit in there again?